### PR TITLE
Snapshots

### PR DESCRIPTION
**Pull request by felipemadero**
Add network state preservation to local/server/client

### CODE REVIEW

File: local/network.go [hunk 1]
```diff
@@ -4,13 +4,17 @@ import (
 	"context"
 	"embed"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io"
 	"io/fs"
+	"io/ioutil"
```
Comment (danlaine) [inline local/network.go:11]: https://pkg.go.dev/io/ioutil
```
As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details.
```

File: network/network.go [hunk 2]
```diff
@@ -7,10 +7,13 @@ import (
 	"github.com/ava-labs/avalanche-network-runner/network/node"
 )
 
+var ErrUndefined = errors.New("undefined network")
 var ErrStopped = errors.New("network stopped")
 
 // Network is an abstraction of an Avalanche network
 type Network interface {
+	// Load network config
+	LoadConfig(context.Context, Config) error
```
Comment (danlaine) [inline network/network.go:16]: It seems like it's require to call this right after the network is created...if that's so, it should just be part of `NewNetwork`. Also, can we please add a more detailed specification?

Comment (felipemadero) [reply network/network.go:16]: After network instance creation, you can also configure and start the network from a given snapshot. So prefered to let network instance creation to do only fields initialization.
Related reasons:
1) it seems more uniform to add the network config input param to the interface, considering that AddNode also takes a node config
2) snapshot load implementation is based on network config load, and preferred to add all snapshot management functions to the interface

File: server/server.go [hunk 3]
```diff
@@ -868,6 +854,176 @@ func (s *server) SendOutboundMessage(ctx context.Context, req *rpcpb.SendOutboun
...
+	// update cluster info non-blocking
+	// the user is expected to poll this latest information
+	// to decide cluster/subnet readiness
+	go func() {
+		zap.L().Info("waiting for local cluster readiness")
+		select {
+		case <-s.closed:
+			return
+		case <-s.network.stopc:
+			// TODO: fix race from shutdown
```
Comment (dasconnor) [inline server/server.go:937]: do we want a ticket for this?

**Review by dasconnor — COMMENTED** [review]
LGTM

**Review by holisticode — COMMENTED** [review]
Great job here.

Some things aren't very clear to me yet, I request a call to go through some questions.
Also, the design as-is here will require that clients (e.g. `avalanche-cli`) have to implement state control themselves client-side, i.e. start the network , save  the snapshot, load the snapshot etc.

You may need to update the README for the changes here

File: server/server.go (hunk 3 above)
Comment (felipemadero) [reply server/server.go:937]: https://ava-labs.atlassian.net/browse/AV-1641
ok added. it is previous to this PR and will need a little research probably.

File: local/network.go [hunk 4]
```diff
@@ -136,6 +148,17 @@ func init() {
 		defaultNetworkConfig.NodeConfigs[i].CChainConfigFile = string(cChainConfig)
 		defaultNetworkConfig.NodeConfigs[i].IsBeacon = true
 	}
+
+	// create default snapshots dir
+	usr, err := user.Current()
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:155]: currently this is in init()

File: local/network.go [hunk 5]
```diff
@@ -136,6 +148,17 @@ func init() {
...
+	}
+	defaultSnapshotsDir = filepath.Join(usr.HomeDir, snapshotsRelPath)
+	err = os.MkdirAll(defaultSnapshotsDir, os.ModePerm)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:160]: currently this is in init()

File: examples/local/fivenodenetwork/main.go [hunk 6]
```diff
@@ -64,10 +64,13 @@ func main() {
 
 func run(log logging.Logger, binaryPath string) error {
 	// Create the network
-	nw, err := local.NewDefaultNetwork(log, binaryPath)
+	nw, err := local.NewNetwork(log, "", "")
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply examples/local/fivenodenetwork/main.go:67]: currently server can be given the snapshotsDir on server instantiation, and rootDir can be given via http call to Start/RestartNode which I shall avoid and made it be given on server instantiation

File: local/network.go [hunk 7]
```diff
@@ -610,6 +660,171 @@ func (ln *localNetwork) removeNode(nodeName string) error {
...
+	for _, nodeConfig := range nodesConfig {
+		sourceDbDir, ok := nodesDbDir[nodeConfig.Name]
+		if !ok {
+			return fmt.Errorf("failure obtaining db path for node %q", nodeConfig.Name)
+		}
+		sourceDbDir = filepath.Join(sourceDbDir, constants.NetworkName(ln.networkID))
+		targetDbDir := filepath.Join(filepath.Join(snapshotDbDir, nodeConfig.Name), constants.NetworkName(ln.networkID))
+		if err := dircopy.Copy(sourceDbDir, targetDbDir); err != nil {
+			return fmt.Errorf("failure saving node %q db dir: %w", nodeConfig.Name, err)
+		}
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:740]: the iteration belong to somewhat different logical steps

File: local/network.go [hunk 8]
```diff
@@ -660,8 +882,14 @@ func addNetworkFlags(log logging.Logger, networkFlags map[string]interface{}, no
 func (ln *localNetwork) setNodeName(nodeConfig *node.Config) error {
 	// If no name was given, use default name pattern
 	if len(nodeConfig.Name) == 0 {
-		nodeConfig.Name = fmt.Sprintf("%s%d", defaultNodeNamePrefix, ln.nextNodeSuffix)
-		ln.nextNodeSuffix++
+		for {
+			nodeConfig.Name = fmt.Sprintf("%s%d", defaultNodeNamePrefix, ln.nextNodeSuffix)
+			ln.nextNodeSuffix++
+			_, ok := ln.nodes[nodeConfig.Name]
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:899]: its ok. can't block forever it searching for a map key and changing it

File: local/network.go [hunk 9]
```diff
@@ -690,13 +918,24 @@ func makeNodeDir(log logging.Logger, rootDir, nodeName string) (string, error) {
...
+func getConfigEntry(
+	nodeConfigFlags map[string]interface{},
+	configFile map[string]interface{},
+	flag string,
+	defaultVal string,
+) (string, error) {
 	var entry string
+	if val, ok := nodeConfigFlags[flag]; ok {
+		if entry, ok := val.(string); ok {
+			return entry, nil
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:941]: node config flags have precedence over config file but that was not reflected in this function

File: local/network.go [hunk 10]
```diff
@@ -710,8 +949,10 @@ func getPort(
 	if portIntf, ok := flags[portKey]; ok {
 		if portFromFlags, ok := portIntf.(int); ok {
 			port = uint16(portFromFlags)
+		} else if portFromFlags, ok := portIntf.(float64); ok {
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:963]: difference in the type depending on the origin of the flags: created by the user in code (int) vs loaded from network config in json (float64)

File: server/network.go [hunk 11]
```diff
@@ -111,84 +110,83 @@ func newLocalNetwork(opts localNetworkOptions) (*localNetwork, error) {
...
+		mergedConfig, err := mergeNodeConfig(defaultConfig, globalConfig, lc.options.customNodeConfigs[nodeName])
 		if err != nil {
-			return nil, fmt.Errorf("failed merging provided configs: %w", err)
+			return fmt.Errorf("failed merging provided configs: %w", err)
 		}
 
-		cfg.NodeConfigs[i].ConfigFile, err = createConfigFileString(mergedConfig, logDir, dbDir, opts.pluginDir, opts.whitelistedSubnets)
-		if err != nil {
-			return nil, err
+		buildDir := ""
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply server/network.go:169]: avalanchego is not having param pluginDir, instead it expects the parent dir of it, called buildDir

File: server/server.go [hunk 12]
```diff
@@ -872,6 +857,177 @@ func (s *server) SendOutboundMessage(ctx context.Context, req *rpcpb.SendOutboun
...
+		case <-s.network.stopCh:
+			return
+		case serr := <-s.network.startErrCh:
+			zap.L().Warn("snapshot load failed to complete", zap.Error(serr))
+			s.mu.Lock()
+			s.network = nil
+			s.clusterInfo = nil
+			s.mu.Unlock()
+			return
+		case <-s.network.customVMsReadyCh:
```
Comment (holisticode) [inline server/server.go:968]: What if there are no custom VMs?

File: local/network.go [hunk 13]
```diff
@@ -183,93 +205,93 @@ func (npc *nodeProcessCreator) NewNodeProcess(config node.Config, args ...string
...
 			stderr:      os.Stderr,
 		},
-		dir,
+		rootDir,
+		snapshotsDir,
 	)
+	if err != nil {
+		return net, err
+	}
+	err = net.loadConfig(context.Background(), networkConfig)
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:233]: that is somewhat difficult to follow for me. strong secondary effects example.

[the remark this answers is not in the document]
Comment (holisticode) [reply local/network.go:233]: Isn't it the same pattern though as

`if err:= ...; err != nil` vs
```
err := ....
if err != nil 
```

File: server/server.go (hunk 12 above)
Comment (holisticode) [reply server/server.go:968]: Hmm. 
1. Shouldn't we check for any existing VMs first at all? Like https://github.com/ava-labs/avalanche-network-runner/blob/58c0e5f1a1ceb3fcf48ced18867330ec8a468f7c/server/server.go#L383, why have to wait for Custom VMs if there are none?
2. Doesn't closing of `customVMsReadyCh` prevent any future custom VMs to be deployed properly?
3. Two subsequent `select` looks odd to me, can we not avoid it ?

File: local/network.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply local/network.go:233]: it is not exactly the same. Here we return an object reference, and the error result of a function that for a secondary effect changes the first returned object. Anyway I can change the code if you think it is more readable.

File: server/server.go (hunk 12 above)
Comment (felipemadero) [reply server/server.go]: Done 3. For 1, LoadSnapshot() is not currently receiving requests to start custom VMs. It is just loading a snapshot. There is not currently on server a helper function to deploy a VM on a already started network. Probably next thing to add to the server.

**Pull request merged**